activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -116,10 +116,11 @@ def concat(*records)
         records = records.flatten
         if owner.new_record?
           load_target
-          concat_records(records)
+          result = concat_records(records)
         else
-          transaction { concat_records(records) }
+          result = transaction { concat_records(records) }
         end
+        result && self
```

Comment:
why the `&&` is being used here? It should always return `self`

Reply:
The transaction above could be rolled back without exception. So this method should notify failure to caller.

```rb
        def concat_records(records, raise = false)
          ...

          raise ActiveRecord::Rollback unless result

          records
        end
```

And this method is called by `replace_records`, which expects falsey value when transaction was rolled back.

```
        def replace_records(new_target, original_target)
          delete(difference(target, new_target))

          unless concat(difference(new_target, target))
            @target = original_target
            raise RecordNotSaved, "Failed to replace #{reflection.name} because one or more of the " \
                                  "new records could not be saved."
          end
```

Note that rollback was introduced in this commit: https://github.com/rails/rails/commit/5dc72378b783e924c5bf079ca660388ec4ac9224#diff-20f545c453ee24942b6f7ae565e9e369R450

Reply:
Ah, maybe `RecordNotSaved` should be raised by `concat` method itself instead of returning falsey value, isn't it..?

Reply:
Remaining the doc style comment is a historical reason, `CollectionAssociation` was replaced by `CollectionProxy` as public API, now `CollectionAssociation` is private API.

https://github.com/rails/rails/blob/272e64101bb6daaa4bbaf2d1c64f64c6b82ead8e/activerecord/lib/active_record/associations/collection_association.rb#L28

Thus we are not always necessary to change the implementation of `CollectionAssociation#concat` unless the change fixes any public behavior.